cmd/whitesourceExecuteScan.go

```diff
@@ -147,11 +149,30 @@ func runWhitesourceExecuteScan(config *ScanOptions, scan *ws.Scan, utils whiteso
 }
 
 func runWhitesourceScan(config *ScanOptions, scan *ws.Scan, utils whitesourceUtils, sys whitesource, commonPipelineEnvironment *whitesourceExecuteScanCommonPipelineEnvironment) error {
+	// Download Docker image for container scan
+	// ToDo: move it to improve testability
+	if config.BuildTool == "docker" {
+		saveImageOptions := containerSaveImageOptions{
+			ContainerImage:       config.ScanImage,
+			ContainerRegistryURL: config.ScanImageRegistryURL,
+			IncludeLayers:        config.ScanImageIncludeLayers,
+		}
+		dClientOptions := piperDocker.ClientOptions{ImageName: saveImageOptions.ContainerImage, RegistryURL: saveImageOptions.ContainerRegistryURL, LocalPath: "", IncludeLayers: saveImageOptions.IncludeLayers}
+		dClient := &piperDocker.Client{}
+		dClient.SetOptions(dClientOptions)
+		if err := runContainerSaveImage(&saveImageOptions, &telemetry.CustomData{}, "./cache", "", dClient); err != nil {
+			return errors.Wrapf(err, "failed to dowload Docker image %v", config.ScanImage)
+		}
+
+	}
+
 	// Start the scan
 	if err := executeScan(config, scan, utils); err != nil {
 		return err
 	}
 
+	// ToDo: Check this:
+	// Why is this required at all, resolveProjectIdentifiers() is already called before the scan in runWhitesourceExecuteScan()
```

Comment:
Initially, the code only considered the "aggregate project" use-case. For example the project token was not yet available if the project is created by the scan itself. That is why the method is called another time after the scan, it seeks to fill out any remaining properties. Just sharing my findings, I don't mean to imply that I like this design. It confused me as well at first.